src/Microsoft.Identity.Client/Platforms/iOS/iOSBroker.cs

```diff
@@ -0,0 +1,236 @@
+        private IServiceBundle _serviceBundle;
+
+        public iOSBroker(ICoreLogger logger)
+        {
+            _logger = logger;
+        }
+
+        public bool CanInvokeBroker(OwnerUiParent uiParent, IServiceBundle serviceBundle)
+        {
+            _serviceBundle = serviceBundle;
+
+            if (uiParent == null)
+            {
+                _logger.Verbose(iOSBrokerConstants.UiParentIsNullCannotInvokeBroker);
+                return false;
+            }
+
+            if (uiParent.CoreUiParent.CallerViewController == null)
+            {
+                _logger.Verbose(iOSBrokerConstants.CallerViewControllerIsNullCannotInvokeBroker);
+                return false;
+            }
+
+            var result = false;
+
+            if (_serviceBundle.Config.IsBrokerEnabled)
+            {
+                _logger.Verbose(iOSBrokerConstants.CanInvokeBroker + _serviceBundle.Config.IsBrokerEnabled);
+
+                uiParent.CoreUiParent.CallerViewController.InvokeOnMainThread(() =>
+                {
+                    result = UIApplication.SharedApplication.CanOpenUrl(new NSUrl("msauthv2://"));
+                });
+            }
+            if (!result)
+            {
+                _logger.Verbose(result + iOSBrokerConstants.CanInvokeBrokerReturnsFalseMessage);
+            }
+
+            return result;
+        }
+
+        public async Task<MsalTokenResponse> AcquireTokenUsingBrokerAsync(Dictionary<string, string> brokerPayload, IServiceBundle serviceBundle)
+        {
+            _brokerPayload = brokerPayload;
+
+            CheckBrokerPayloadForSilentFlow();
+
+            AddIosSpecificParametersToPayload();           
+
+            await InvokeIosBrokerAsync().ConfigureAwait(false);            
+
+            return ProcessBrokerResponse();
+        }
+
+        private void CheckBrokerPayloadForSilentFlow()
+        {
+            if (_brokerPayload.ContainsKey(BrokerParameter.SilentBrokerFlow))
+            {
+                throw new MsalUiRequiredException(MsalError.FailedToAcquireTokenSilently, MsalErrorMessage.FailedToAcquireTokenSilently);
+            }
+        }
+
+        private void AddIosSpecificParametersToPayload()
+        {
+            string base64EncodedString = Base64UrlHelpers.Encode(BrokerKeyHelper.GetRawBrokerKey(_logger));
+            _brokerPayload[iOSBrokerConstants.BrokerKey] = base64EncodedString;
+            _brokerPayload[iOSBrokerConstants.MsgProtocolVer] = "3";
+
+            if (_brokerPayload.ContainsKey(iOSBrokerConstants.Claims))
+            {
+                _brokerPayload.Add(iOSBrokerConstants.SkipCache, "YES");
+                string claims = Base64UrlHelpers.Encode(_brokerPayload[BrokerParameter.Claims]);
+                _brokerPayload[BrokerParameter.Claims] = claims;
+            }
+        }
+
+        private async Task InvokeIosBrokerAsync()
+        {
+            brokerResponse = null;
+            brokerResponseReady = new SemaphoreSlim(0);
```

Comment:
>brokerResponseReady = new SemaphoreSlim(0); [](start = 12, length = 43)

move under if

Reply:
can't because i need it for the else, which does get it. probably easier to show you.

---
In reply to: [256627943](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/881#discussion_r256627943) [](ancestors = 256627943)

Reply:
#!